File: backend/app/schemas/photo.py
"""照片上传相关 Schema"""

from pydantic import BaseModel, Field


class StsCredentialOut(BaseModel):
    """STS 临时上传凭证（与前端 StsCredential 对齐）"""

    tmp_secret_id: str
    tmp_secret_key: str
    session_token: str
    expired_time: int = Field(..., description="过期时间戳（秒）")
    start_time: int = Field(..., description="起始时间戳（秒）")
    bucket: str
    region: str
    upload_host: str
    upload_dir: str
    file_path: str = Field(..., description="本次上传的完整对象 key 前缀")
    url_prefix: str = Field(..., description="拼接访问 URL 的前缀")
    # 预签名 URL：前端直接 PUT 文件到此 URL，无需计算签名
    presigned_url: str = Field("", description="COS 预签名 PUT URL")
    object_key: str = Field("", description="本次上传的 COS object key")


class PhotoConfirmReq(BaseModel):
    """上传确认请求"""

    file_id: str = Field(..., description="前端生成的文件唯一 ID")
    url: str = Field(..., description="上传后的对象 key 或完整 URL")
    size: int = Field(0, description="文件大小（字节）")
    width: int | None = None
    height: int | None = None


class UploadedPhotoOut(BaseModel):
    """上传确认返回（与前端 UploadedPhoto 对齐）"""

    file_id: str
    url: str
    thumbnail_url: str
    size: int
    width: int = 0
    height: int = 0
    photo_id: str = Field("", description="后端 Photo 记录 ID")

File: travel-photo-miniprogram/src/api/photo.ts
/**
 * 照片上传API
 */
import { request } from './request'

export interface StsCredential {
  tmp_secret_id: string
  tmp_secret_key: string
  session_token: string
  expired_time: number
  start_time: number
  bucket: string
  region: string
  upload_host: string
  upload_dir: string
  file_path: string
  url_prefix: string
  presigned_url: string
  object_key: string
}

export interface UploadedPhoto {
  file_id: string
  url: string
  thumbnail_url: string
  size: number
  width: number
  height: number
}

/**
 * 申请上传凭证（STS）
 */
export function getUploadCredential(fileCount: number): Promise<StsCredential> {
  return request<StsCredential>({
    url: `/api/v1/photos/sts?file_count=${fileCount}`,
    method: 'GET',
  })
}

/**
 * 获取预签名上传 URL
 */
export function getPresignedUrl(objectKey: string): Promise<{
  presigned_url: string
  object_key: string
  access_url: string
}> {
  return request({
    url: '/api/v1/photos/presign',
    method: 'POST',
    data: { object_key: objectKey },
  })
}

/**
 * 上传完成后回传文件信息
 */
export function confirmUpload(payload: {
  file_id: string
  url: string
  size: number
  width?: number
  height?: number
}): Promise<UploadedPhoto> {
  return request<UploadedPhoto>({
    url: '/api/v1/photos/confirm',
    method: 'POST',
    data: payload,
  })
}

/**
 * 确认下载（保存）精修图
 *
 * 用户把精修图保存到相册后调用，后端据此立即删除 COS 中的原图，
 * 以支持「先看原图/精修图对比，再清理原图」的存储策略。幂等。
 */
export function confirmDownload(photoId: string): Promise<{
  photo_id: string
  original_deleted: boolean
}> {
  return request({
    url: `/api/v1/photos/${photoId}/download`,
    method: 'POST',
  })
}
